crates/file-serve/src/lib.rs

```diff
@@ -100,17 +105,45 @@ impl Server {
         self.addr.as_str()
     }
 
+    /// Whether the server is running
+    pub fn is_running(&self) -> bool {
+        self.running.load(Ordering::Acquire)
+    }
+
     /// Start the webserver
-    pub fn serve(&self) -> Result<std::convert::Infallible, Error> {
+    pub fn serve(&self) -> Result<(), Error> {
+        if self.is_running() {
+            return Err(Error::new("the server is running"));
+        }
+
         // attempts to create a server
         let server = tiny_http::Server::http(self.addr()).map_err(Error::new)?;
-
-        for request in server.incoming_requests() {
-            if let Err(e) = static_file_handler(self.source(), request) {
-                log::error!("{}", e);
+        self.running.store(true, Ordering::Release);
```

Comment:
Shouldn't these use a compare-exchange?

Technically, the new API allows for multiple threads can call `serve` and there will be a race that would cause multiple to get past the first check (and then fail when they try to bind).